qiskit/quantum_info/synthesis/two_qubit_decompose.py

```diff
@@ -0,0 +1,37 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2017, 2019.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+# pylint: disable=unused-import
+
+"""
+Expand 2-qubit Unitary operators into an equivalent
+decomposition over SU(2)+fixed 2q basis gate, using the KAK method.
+
+May be exact or approximate expansion. In either case uses the minimal
+number of basis applications.
+
+Method is described in Appendix B of Cross, A. W., Bishop, L. S., Sheldon, S., Nation, P. D. &
+Gambetta, J. M. Validating quantum computers using randomized model circuits.
+arXiv:1811.12926 [quant-ph] (2018).
+"""
+
+from __future__ import annotations
+import warnings
+
+from qiskit.synthesis.two_qubit.two_qubit_decompose import TwoQubitWeylDecomposition
```

Comment:
There are a few public decomposer classes in the module. It's probably better to play it safe with:

```suggestion
from qiskit.synthesis.two_qubit.two_qubit_decompose import *
```

and expose them all.

Reply:
When I try to do this change, I get the following pylint errors:


```
qiskit/quantum_info/synthesis/two_qubit_decompose.py:30:0: W0401: Wildcard import qiskit.synthesis.two_qubit.two_qubit_decompose (wildcard-import)
qiskit/quantum_info/synthesis/two_qubit_decompose.py:30:0: W0614: Unused import(s) cmath, math, io, base64, logging, np, logger, decompose_two_qubit_product_gate, TwoQubitWeylDecomposition, TwoQubitWeylIdEquiv, TwoQubitWeylSWAPEquiv, TwoQubitWeylPartialSWAPEquiv, TwoQubitWeylPartialSWAPFlipEquiv, TwoQubitWeylControlledEquiv, TwoQubitControlledUDecomposer, TwoQubitWeylMirrorControlledEquiv, TwoQubitWeylfSimaabEquiv, TwoQubitWeylfSimabbEquiv, TwoQubitWeylfSimabmbEquiv, TwoQubitWeylGeneral, Ud, trace_to_fid, rz_array, TwoQubitBasisDecomposer, TwoQubitDecomposeUpToDiagonal, two_qubit_cnot_decompose, ClassVar, Optional, Type, QuantumRegister, QuantumCircuit, Gate, CXGate, RXGate, RYGate, RZGate, QiskitError, Operator, weyl_coordinates, transform_to_magic_basis, OneQubitEulerDecomposer, DEFAULT_ATOL and deprecate_arg from wildcard import of qiskit.synthesis.two_qubit.two_qubit_decompose (unused-wildcard-import)

```

Reply:
I can add more classes to the import (and later also to the main branch documentation), but I think it would be better to state them specifically

Reply:
Note that except of `TwoQubitWeylDecomposition, TwoQubitControlledUDecomposer, TwoQubitDecomposeUpToDiagonal`, all other classes are variants of `TwoQubitWeylDecomposition`.

Reply:
A search that I did in various libraries (`circuit knitting`, `qiskit-ibm-runtime` and `qiskit-ibm-provider`) does not indicate the usage of other classes except of `TwoQubitWeylDecomposition`.

Reply:
I was mostly worried about other downstream users that we don't know about, or that don't have public code anywhere that we can access. I think listing all the public classes explicitly is fine, being explicit doesn't hurt anything. It'll fix the pylint errors too.

Reply:
OK, I added all the classes in the file `two_qubit_decompose`

Reply:
> I was mostly worried about other downstream users that we don't know about

Though there's loads of other specific imports that are still going to break. Here we're only fixing the two-qubit one, but before (<=0.45) people were able to import e.g. `qiskit.quantum_info.synthesis.one_qubit_decompose`, `.qsd`, `.weyl`, `.local_invariance`....

Reply:
That's a fair point, on the one hand we're documenting this release as "fully compatible with the Qiskit 0.45.x releases" but with extra deprecations. But we've also documented the public interface as those that are documented. This is a weird case because `qiskit.quantum_info.synthesis` is a dark corner of the API where there was a ton of undocumented functionality that wasn't exposed anywhere except in the leaf modules, and the definition of the stable api only really changed after 0.45 (in preparation for 1.0.0).

I'm inclined to just err on the side of caution more here just so people don't have an excuse for not upgrading to 0.46 and add module shims back in for everything that redirects to the new location in`qiskit.synthesis` with a deprecation warning. (like what's done for `__init__.py` here already)